docs/src/developer/design/user_api/add_target.rst

```diff
@@ -0,0 +1,309 @@
+.. See the License for the specific language governing permissions and
+.. limitations under the License.
+
+.. _designing_cmaizes_add_target_functions:
+
+#######################################
+Designing CMaize's Add Target Functions
+#######################################
+
+The discussion on the :ref:`designing_cmaizes_user_api` page motivated the need
+for user-facing functions for declaring and adding targets to the build system.
+
+*****************************************
+What Are the CMaize Add Target Functions?
+*****************************************
+
+CMaize's user-facing :term:`API` is comprised of functions designed to
+integrate easily into CMake-style build systems. CMaize's add target functions
+are the pieces of the user :term:`API` through which users can define
+:term:`build targets<build target>`, i.e., the libraries, executables, etc.
+comprising their :term:`project`. We refer to these functions as the "add
+target" functions for generality.
+
+***********************************************
+Why Do We Need The CMaize Add Target Functions?
+***********************************************
+
+Users need a way to declare :term:`build targets<build target>`. Traditional
+CMake provides only two add target functions:
+
+- `add_library <https://cmake.org/cmake/help/latest/command/add_library.html>`_
+- `add_executable <https://tinyurl.com/4pxh3cmf>`_
+
+though it should be noted the `add_library`_ actually covers a number of use
+cases, i.e., static libraries, shared libraries, interface libraries, modules,
+and object libraries. Modern CMake is target-based and CMaize will ultimately
+need to make targets in order to interact with CMake. Thus CMaize's add target
+functions will necessarily wrap CMake's add target functions.
+
+The primary reason CMaize needs wrapper functions is to capture the user's
+input. More specifically, the CMaize versions of the add target functions will
+record the target options in the active ``CMaizeProject`` so that CMaize will
+be able to automate packaging of the project. Another motivation for
+maintaining a separate set of CMaize add target functions is to
+provide more succinct APIs. In practice, declaring CMake targets can be very
+verbose and much of what needs to be provided can be inferred or learned from
+other sources.
+
+**********************
+Add Target Terminology
+**********************
+
+CMaize will be dealing with targets associated with multiple coding languages.
+Most coding languages organize code similarly, but differ in what they term
+those organizational units. In order to provide a unified description we define
+the following terms. We use these terms throughout this page regardless of what
+the coding-language appropriate terms are.
+
+.. glossary::
+
+   executable
+      A program meant to be run by a user. It may be compiled or it may not be,
+      e.g., a Python script with a ``main`` function.
```

Comment:
```suggestion
      e.g., a Python script.
```

Technically the Python script does not need a `main` function to be run through something like `python <script_name>.py`. Although I think a `main()` function is useful, you could just write a few lines of procedural code in a file and then execute the file, right?